### PR TITLE
[Perf][Benchmark] Add multi-dim logsumexp benchmarks

### LINKED ISSUE #900: [PERF][Benchmark] Add logsumexp to multi-dim reduction benchmarks
# [PERF][Benchmark] Add logsumexp to multi-dim reduction benchmarks

### Symptom / Motivation
Discovered during PR #894 ([Perf][Benchmark] Add multi-dim reduction benchmarks).

Issue #822 listed five benchmark file groups for multi-dim coverage:
1. `bench_reduce.py` — sum, mean, amax, amin, prod, std, var, var_mean
2. `bench_argreduce.py` — argmax, argmin
3. `bench_logical_reduce.py` — all, any, count_nonzero
4. `bench_vector_norm.py` — l1_norm, l2_norm, inf_norm
5. `bench_softmax.py` — **logsumexp**

PR #894 added multi-dim benchmarks for groups 1-4 but omitted group 5 (logsumexp). The file `bench_reduce_multidim.py` does not contain any logsumexp/softmax cases.

### Root Cause Analysis
PR #894 scope grew during dev/review iterations (argreduce kernel limitations, skip handler cleanup). Logsumexp was deprioritized and not added before the PR was finalized.

### Related Files
- `benchmarks/ops/bench_softmax.py` — existing single-dim logsumexp benchmarks
- `benchmarks/ops/bench_reduce_multidim.py` — target file for multi-dim cases
- `tileops/ops/reduction/` — logsumexp kernel implementation

## Goal
Add multi-dim logsumexp benchmark cases to complete the five-group coverage from issue #822.

## Plan

1. Add a logsumexp section to `bench_reduce_multidim.py` (or `bench_softmax.py`) with multi-dim parametrizations (e.g., 3D tensor with `dim=[0, 2]`).
2. Include `keepdim=True` and `keepdim=False` variants.
3. Verify benchmarks run without error.

## Constraints
- Must not regress PR #894
- Follow existing `WorkloadBase` / `BenchmarkBase` patterns in `bench_reduce_multidim.py`

## Acceptance Criteria
- [ ] AC-1: At least one multi-dim logsumexp benchmark case exists
- [ ] AC-2: Benchmark runs without error on target GPU
- [ ] AC-3: `pytest --collect-only` collects the new cases without errors

### PR DESCRIPTION
**Pull request by lcy-seso**
## Summary

Add multi-dim logsumexp benchmark cases to `bench_reduce_multidim.py`, completing the five-group reduction coverage from issue #822. Six new parametrized cases cover `dim=[0, 2]` and `dim=[0, 1]` on 3D tensors with `keepdim=True/False` and `float16/bfloat16` variants.

Closes #900

## Test plan

- [x] **AC-1**: At least one multi-dim logsumexp benchmark case exists
  - pytest --collect-only shows 6 logsumexp multi-dim cases: lse-7B-dim02-nokeepdim, lse-7B-dim02-keepdim, lse-7B-dim01-nokeepdim, lse-7B-dim01-keepdim-bf16, lse-7B-longctx-dim02, lse-7B-longctx-dim02-keepdim-bf16. All use multi-dim (dim=[0,2] or dim=[0,1]) on 3D tensors. 27 tests collected total.
- [x] **AC-2**: Benchmark runs without error on target GPU
  - pytest -k logsumexp -v: 6 passed, 21 deselected in 6.31s. Full suite: 27 passed in 9.96s. No errors or failures.
- [x] **AC-3**: pytest --collect-only collects the new cases without errors
  - pytest --collect-only: 27 tests collected in 0.02s. All 6 test_logsumexp_multidim_bench cases appear in output with correct parametrized IDs.

## Validation

- **Tests**: 27 total, 27 passed, 0 failed
- **GPU**: NVIDIA H200, all 6 logsumexp benchmarks pass

## Follow-up

No follow-up issues. Suggestions: `bench_softmax.py` runs `LogSumExpFwdOp` with `tune=True` while `bench_reduce_multidim.py` uses the default `tune=False` for all 6 groups — if cross-file performance comparison is needed, consider adding a `tune=True` variant.

### CODE REVIEW

File: benchmarks/ops/bench_reduce_multidim.py [hunk 1]
```diff
@@ -602,5 +602,123 @@ def test_cumulative_multidim_bench(
...
+
+
+class LogSumExpMultidimBenchmark(BenchmarkBase):
+    def calculate_flops(self) -> Optional[float]:
+        t = self.workload
+        total_elems = 1
+        for s in t.shape:
+            total_elems *= s
+        # logsumexp: exp per element + sum + log ≈ 3 * total_elems
+        return 3 * total_elems
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply benchmarks/ops/bench_reduce_multidim.py]: Addressed in dfd521d. Changed calculate_flops to return total_elems instead of 3 * total_elems, matching the convention used by all other reduction benchmarks in this file. Current HEAD: dfd521d

File: benchmarks/ops/bench_reduce_multidim.py [hunk 2]
```diff
@@ -602,5 +602,123 @@ def test_cumulative_multidim_bench(
...
+
+    def calculate_memory(self) -> Optional[float]:
+        t = self.workload
+        elem_bytes = torch.tensor([], dtype=t.dtype).element_size()
+        total_elems = 1
+        for s in t.shape:
+            total_elems *= s
+        out_elems = 1
+        dims = set(d % len(t.shape) for d in t.dim)
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply benchmarks/ops/bench_reduce_multidim.py:690]: This concern does not apply to the current benchmark. All fixture params pass dim as a list:

```python
# From LogSumExpMultidimFixture.PARAMS (lines 611-646):
(4, 128, 4096), [0, 2], False, torch.float16    # list
(4, 128, 4096), [0, 2], True, torch.float16     # list
(4, 128, 4096), [0, 1], False, torch.float16    # list
(4, 128, 4096), [0, 1], True, torch.bfloat16    # list
(2, 512, 4096), [0, 2], False, torch.float16    # list
(2, 512, 4096), [0, 2], True, torch.bfloat16    # list
```

The type hint is `dim: list` and the `LogSumExpMultidimTest.__init__` signature enforces `dim: list`. No single-integer dim path exists in these multi-dim benchmarks by design.

Current HEAD: dfd521d

File: benchmarks/ops/bench_reduce_multidim.py [hunk 3]
```diff
@@ -602,5 +602,122 @@ def test_cumulative_multidim_bench(
...
+    keepdim: bool,
+    dtype: torch.dtype,
+) -> None:
+    test = LogSumExpMultidimTest(shape, dim, keepdim, dtype)
+    bm = LogSumExpMultidimBenchmark(test)
+    inputs = test.gen_inputs()
+
+    op = _make_logsumexp_op(dtype, dim, keepdim)
+    result = bm.profile(op, *inputs)
+    BenchmarkReport.record(op, locals(), result, tag="tileops")
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply benchmarks/ops/bench_reduce_multidim.py:716]: This is a false positive. The file-local convention in bench_reduce_multidim.py is tune=False for ALL benchmark groups -- none of the 5 existing groups (reduce, argreduce, logical_reduce, vector_norm, cumulative) use tune=True. The new logsumexp section correctly follows this file-local pattern.

The comparison to bench_softmax.py is inapplicable: that file benchmarks 1D reductions with different tuning needs. Multi-dim reduction kernels may not have tuning configs available.

Current HEAD: dfd521d

**Review by Ibuki-wind — APPROVED** [review]
Approved at current head dfd521d. Benchmark-layer only, trust-boundary clean, and the added multi-dim logsumexp cases follow the existing tileops/torch multidim benchmark pattern without introducing review findings.
